File: app/src/main/java/com/mikhailovskii/trakttv/TraktTvApp.java
package com.mikhailovskii.trakttv;

import android.app.Application;
import android.content.Context;

import androidx.annotation.NonNull;
import androidx.room.Room;

import com.facebook.stetho.Stetho;
import com.mikhailovskii.trakttv.db.room.MovieDatabase;

public class TraktTvApp extends Application {

    public static TraktTvApp instance;

    private static String DB_NAME = "MovieDatabase";
    private static Context sAppContext;
    private MovieDatabase database;

    @NonNull
    public static Context getAppContext() {
        return sAppContext;
    }

    public static TraktTvApp getInstance() {
        return instance;
    }

    @Override
    public void onCreate() {
        super.onCreate();
        sAppContext = getApplicationContext();
        instance = this;
        database = Room.databaseBuilder(this, MovieDatabase.class, DB_NAME)
                .fallbackToDestructiveMigration()
                .build();
        setupDebugTools();
    }

    public MovieDatabase getDatabase() {
        return database;
    }

    private void setupDebugTools() {
        if (BuildConfig.DEBUG) {
            Stetho.initializeWithDefaults(this);
        }
    }

}

File: app/src/main/java/com/mikhailovskii/trakttv/TraktTvApp.kt
package com.mikhailovskii.trakttv

import android.app.Application
import android.content.Context
import com.crashlytics.android.Crashlytics
import com.crashlytics.android.ndk.CrashlyticsNdk
import com.facebook.stetho.Stetho
import com.google.android.gms.tasks.OnCompleteListener
import com.google.firebase.iid.FirebaseInstanceId
import com.mikhailovskii.trakttv.di.AppModule.apiModule
import com.mikhailovskii.trakttv.di.AppModule.appModule
import com.mikhailovskii.trakttv.di.AppModule.dbModule
import com.mikhailovskii.trakttv.di.AppModule.mvpModule
import io.fabric.sdk.android.Fabric
import org.koin.android.ext.koin.androidContext
import org.koin.android.ext.koin.androidLogger
import org.koin.core.context.startKoin
import timber.log.Timber

class TraktTvApp : Application() {

    override fun onCreate() {
        super.onCreate()

        appContext = applicationContext

        instance = this

        initStetho()
        initFabric()

        Timber.plant(Timber.DebugTree())

        startKoin {
            androidLogger()
            androidContext(this@TraktTvApp)
            modules(appModule, apiModule, dbModule, mvpModule)
        }

        FirebaseInstanceId.getInstance().instanceId
                .addOnCompleteListener(OnCompleteListener { task ->

                    if (!task.isSuccessful) {
                        return@OnCompleteListener
                    }

                })
    }

    private fun initFabric() {
        Fabric.with(this, Crashlytics(), CrashlyticsNdk())
    }

    private fun initStetho() {
        if (BuildConfig.DEBUG) {
            Stetho.initializeWithDefaults(this)
        }
    }


    companion object {

        lateinit var instance: TraktTvApp
        lateinit var appContext: Context

    }

}
